common/src/main/java/org/apache/nemo/common/DirectByteBufferOutputStream.java

```diff
@@ -0,0 +1,163 @@
+    dataList.addLast(ByteBuffer.allocateDirect(pageSize));
+  }
+
+  /**
+   * Writes the specified byte to this output stream.
+   *
+   * @param   b   the byte to be written.
+   */
+  @Override
+  public void write(final int b) {
+    ByteBuffer currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
+    if (currentBuf == null || currentBuf.remaining() <= 0) {
+      newLastBuffer();
+      currentBuf = dataList.getLast();
+    }
+    currentBuf.put((byte) b);
+  }
+
+  /**
+   * Writes {@code b.length} bytes from the specified byte array
+   * to this output stream.
+   *
+   * @param b the byte to be written.
+   */
+  @Override
+  public void write(final byte[] b) {
+    write(b, 0, b.length);
+  }
+
+  /**
+   * Writes {@code len} bytes from the specified byte array
+   * starting at offset {@code off} to this output stream.
+   *
+   * @param   b     the data.
+   * @param   off   the start offset in the data.
+   * @param   len   the number of bytes to write.
+   */
+  @Override
+  public void write(final byte[] b, final int off, final int len) {
+    int byteToWrite = len;
+    int offset = off;
+    ByteBuffer currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
+    while (byteToWrite > 0) {
+      if (currentBuf == null || currentBuf.remaining() <= 0) {
+        newLastBuffer();
+        currentBuf = dataList.getLast();
+      }
+      final int bufRemaining = currentBuf.remaining();
+      if (bufRemaining < byteToWrite) {
+        currentBuf.put(b, offset, bufRemaining);
+        offset += bufRemaining;
+        byteToWrite -= bufRemaining;
+      } else {
+        currentBuf.put(b, offset, byteToWrite);
+        offset += byteToWrite;
+        byteToWrite = 0;
+      }
+    }
+  }
+
+  /**
+   * Creates a byte array that contains the whole content
+   * currently written in this output stream.
+   *
+   * @return the current contents of this output stream, as byte array.
+   */
+  public byte[] toByteArray() {
+    if (dataList.isEmpty()) {
+      final byte[] byteArray = new byte[0];
+      return byteArray;
+    }
+
+    ByteBuffer lastBuf = dataList.getLast();
+    // pageSize equals the size of the data filled in the ByteBuffers
+    // except for the last ByteBuffer. The size of the data in the
+    // ByteBuffer can be obtained by calling ByteBuffer.position().
+    final int arraySize = pageSize * (dataList.size() - 1) + lastBuf.position();
+    final byte[] byteArray = new byte[arraySize];
+    int start = 0;
+    int byteToWrite;
+
+    for (ByteBuffer temp : dataList) {
+      // ByteBuffer has to be shifted to read mode by calling ByteBuffer.flip(),
+      // which sets limit to the current position and sets the position to 0.
+      // Note that capacity remains unchanged.
+      temp.flip();
+      byteToWrite = temp.remaining();
+      temp.get(byteArray, start, byteToWrite);
+      start += byteToWrite;
+    }
+    // The limit of the last buffer has to be set to the capacity for re-write.
```

Comment:
I guess the term "additional writing" is better than "re-write", considering that no data is overwritten.